compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -2366,8 +2372,22 @@ public void visit(BLangBinaryExpr binaryExpr) {
         // Set error type as the actual type.
         BType actualType = symTable.semanticError;
 
-        // Look up operator symbol if both rhs and lhs types are error types
-        if (lhsType != symTable.semanticError && rhsType != symTable.semanticError) {
+        // Do not lookup operator symbol for xml sequence additions
+        if (binaryExpr.opKind == OperatorKind.ADD &&
+                (lhsType.tag == TypeTags.XML && TypeTags.isXMLNonSequenceType(rhsType.tag))) {
+            actualType = new BXMLType(TypeTags.XML, BUnionType.create(null, ((BXMLType) lhsType).constraint, rhsType),
+                    false,
+                    null);
+        } else if (binaryExpr.opKind == OperatorKind.ADD &&
+                (rhsType.tag == TypeTags.XML && TypeTags.isXMLNonSequenceType(lhsType.tag))) {
+            actualType = new BXMLType(TypeTags.XML, BUnionType.create(null, ((BXMLType) rhsType).constraint, lhsType),
+                    false, null);
+        } else if (binaryExpr.opKind == OperatorKind.ADD &&
```

Comment:
Why is the XML type special cased? Also, shall we add these second conditions as nested if conditions? i.e., check if it's an ADD op once

Reply:
This was done to overcome an issue which was discussed in https://ballerina-platform.slack.com/archives/CAXS256RX/p1585205684025700

+1 to check ADD operation once, will change that.

Reply:
About nesting the ADD operation. 
We still want to visit the else if section following for add operations witch are non-xml.

If we nest like u suggested, we will have to keep a separate boolean (xml special case is visited - setting its value under each condition) and then check the negation of that boolean on following section. 

I think it would make the code bit unclear. WDYT?

Reply:
Don't we have to handle something like lhs=comment, rhs=procInstruction

We can use a switch statement with fall through like:
```java
        //noinspection SwitchStatementWithTooFewBranches
        switch (binaryExpr.opKind) {
            // Do not lookup operator symbol for xml sequence additions
            case ADD:
                BType l = getXMLConstituents(lhsType);
                BType r = getXMLConstituents(rhsType);

                if (l != null && r != null) {
                    actualType = new BXMLType(TypeTags.XML, BUnionType.create(null, l, r), false, null);
                    break;
                }
                // Fall through
            default:
                if (lhsType != symTable.semanticError && rhsType != symTable.semanticError) {
                    // Look up operator symbol if both rhs and lhs types aren't error or xml types
                    BSymbol opSymbol = symResolver.resolveBinaryOperator(binaryExpr.opKind, lhsType, rhsType);

                    if (opSymbol == symTable.notFoundSymbol) {
                        opSymbol = symResolver.getBinaryEqualityForTypeSets(binaryExpr.opKind, lhsType, rhsType, binaryExpr);
                    }

                    if (opSymbol == symTable.notFoundSymbol) {
                        dlog.error(binaryExpr.pos, DiagnosticCode.BINARY_OP_INCOMPATIBLE_TYPES, binaryExpr.opKind,
                                lhsType, rhsType);
                    } else {
                        if ((binaryExpr.opKind == OperatorKind.EQUAL || binaryExpr.opKind == OperatorKind.NOT_EQUAL) &&
                                (couldHoldTableValues(lhsType, new ArrayList<>()) &&
                                        couldHoldTableValues(rhsType, new ArrayList<>()))) {
                            dlog.error(binaryExpr.pos, DiagnosticCode.EQUALITY_NOT_YET_SUPPORTED, TABLE_TNAME);
                        }

                        binaryExpr.opSymbol = (BOperatorSymbol) opSymbol;
                        actualType = opSymbol.type.getReturnType();
                    }
                }
        }
```

```java

    private BType getXMLConstituents(BType type) {
        BType c = null;
        if (type.tag == TypeTags.XML) {
            c = ((BXMLType) type).constraint;
        } else if (TypeTags.isXMLNonSequenceType(type.tag)) {
            c = type;
        }
        return c;
    }
```

But there is a catch to this kind of solutions, this may confuse someone who is not familiar with fallthrough.

Reply:
+1 added your suggestion. 
I have also removed redundant xml operations from `SymbolTable` (for element + comment like scenarios) since this covers everything.